spark/src/main/scala/org/apache/spark/sql/delta/perf/DeltaOptimizedWriterExec.scala

```diff
@@ -0,0 +1,323 @@
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+import org.apache.spark.sql.delta.util.BinPackingUtils
+
+// scalastyle:off import.ordering.noEmptyLine
+import org.apache.spark._
+import org.apache.spark.internal.config
+import org.apache.spark.internal.config.ConfigEntry
+import org.apache.spark.network.util.ByteUnit
+import org.apache.spark.rdd.RDD
+import org.apache.spark.shuffle._
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.physical.HashPartitioning
+import org.apache.spark.sql.execution.{ShuffledRowRDD, SparkPlan, UnaryExecNode}
+import org.apache.spark.sql.execution.exchange.ShuffleExchangeExec
+import org.apache.spark.sql.execution.metric.{SQLMetric, SQLMetrics, SQLShuffleReadMetricsReporter, SQLShuffleWriteMetricsReporter}
+import org.apache.spark.storage._
+import org.apache.spark.util.ThreadUtils
+
+
+/**
+ * An execution node which shuffles data to a target output of `DELTA_OPTIMIZE_WRITE_SHUFFLE_BLOCKS`
+ * blocks, hash partitioned on the table partition columns. We group all blocks by their
+ * reducer_id's and bin-pack into `DELTA_OPTIMIZE_WRITE_BIN_SIZE` bins. Then we launch a Spark task
+ * per bin to write out a single file for each bin.
+ *
+ * @param child The execution plan
+ * @param partitionColumns The partition columns of the table. Used for hash partitioning the write
+ * @param deltaLog The DeltaLog for the table. Used for logging only
+ */
+case class DeltaOptimizedWriterExec(
+    child: SparkPlan,
+    partitionColumns: Seq[String],
+    @transient deltaLog: DeltaLog
+  ) extends UnaryExecNode with DeltaLogging {
+
+  override def output: Seq[Attribute] = child.output
+
+  private lazy val writeMetrics =
+    SQLShuffleWriteMetricsReporter.createShuffleWriteMetrics(sparkContext)
+  private lazy val readMetrics =
+    SQLShuffleReadMetricsReporter.createShuffleReadMetrics(sparkContext)
+  override lazy val metrics: Map[String, SQLMetric] = Map(
+    "dataSize" -> SQLMetrics.createSizeMetric(sparkContext, "data size")
+  ) ++ readMetrics ++ writeMetrics
+
+  private lazy val childNumPartitions = child.execute().getNumPartitions
+
+  private lazy val numPartitions: Int = {
+    val targetShuffleBlocks = getConf(DeltaSQLConf.DELTA_OPTIMIZE_WRITE_SHUFFLE_BLOCKS)
+    math.min(
+      math.max(targetShuffleBlocks / childNumPartitions, 1),
+      getConf(DeltaSQLConf.DELTA_OPTIMIZE_WRITE_MAX_SHUFFLE_PARTITIONS))
+  }
+
+  @transient private var cachedShuffleRDD: ShuffledRowRDD = _
+
+  @transient private lazy val mapTracker = SparkEnv.get.mapOutputTracker
+
+  /** Creates a ShuffledRowRDD for facilitating the shuffle in the map side. */
+  private def getShuffleRDD: ShuffledRowRDD = {
+    if (cachedShuffleRDD == null) {
+      val resolver = org.apache.spark.sql.catalyst.analysis.caseInsensitiveResolution
+      val saltedPartitioning = HashPartitioning(
```

Comment:
Is there a reason for choosing HashPartitioning over other schemes ?

Reply:
This matches table partition. For unpartitioned tables, it shouldn't really matter that much. But HashPartitioning should be more robust to certain failure cases, e.g. https://issues.apache.org/jira/browse/SPARK-38388.